docs/development/logging.md

```diff
@@ -0,0 +1,165 @@
+Hence, the following sections focus on logging in the controller and admission components only.
+
+## Structured Logging
+
+Similar to [efforts in the Kubernetes project](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md), we want to migrate our component logs to structured logging.
+As motivated above, we will use the logr interface instead of klog though.
+
+You can read more about the motivation behind structured logging in [logr's background and FAQ](https://github.com/go-logr/logr#background) (also see [this blog post by Dave Cheney](http://dave.cheney.net/2015/11/05/lets-talk-about-logging)).
+Also, make sure to check out controller-runtime's [logging guideline](https://github.com/kubernetes-sigs/controller-runtime/blob/v0.11.0/TMP-LOGGING.md) with specifics for projects using the library.
+The following sections will focus on the most important takeaways from those guidelines and give general instructions on how to apply them to Gardener and its controller-runtime components.
+Note: some parts in this guideline differ slightly from controller-runtime's document.
+
+### TL;DR of Structured Logging
+
+❌ stop using `printf`-style logging:
+```go
+var logger *logrus.Logger
+logger.Infof("scaling deployment %s/%s to %d replicas", deployment.Namespace, deployment.Name, replicaCount)
+```
+
+✅ instead, write static log messages and enrich them with additional structured information in form of key-value pairs:
+
+```go
+var logger logr.Logger
+logger.Info("scaling deployment", "deployment", client.ObjectKeyFromObject(deployment), "replicas", replicaCount)
+```
+
+## Log Configuration
+
+Gardener components can be configured to either log in `json` (default) or `text` format:
+`json` format is supposed to be used in production, while `text` format might be nicer for development.
+
+```text
+# json
+{"level":"info","ts":"2021-12-16T08:32:21.059+0100","msg":"hello botanist","garden":"eden"}
+
+# text
+2021-12-16T08:32:21.059+0100    INFO    hello botanist  {"garden": "eden"}
+```
+
+Components can be set to one of the following log levels (with increasing verbosity): `error`, `info` (default), `debug`.
+
+> ℹ️ Note: some Gardener components don't feature a configurable log level and format yet.
+> In this case, they log at `info` in `json` format.
+> We might add configuration options via command line flags that can be used in all components in the future though.
+
+## Log Levels
+
+logr uses [V-levels](https://github.com/go-logr/logr#why-v-levels) (numbered log levels), higher V-level means higher verbosity.
+V-levels are relative (in contrast to `klog`'s absolute V-levels), i.e., `V(1)` creates a logger, that is one level more verbose than its parent loggger.
+
+In Gardener components, `V(0)` of the root logger maps to `info` level, `V(1)` means `debug` level.
```

Comment:
That would be the one constructed in some `main` func or similar (maybe from the component config). `root` because it's not "derived" from some other logger via `V()` or `WithValues()`, meaning it has the original *relative* log level of `0`.

Reply:
Thanks for explaining. Does it makes sense to add a short explanation in parentheses? At least to me, it was not clear before reading your explanation. If you say that you rather prefer to keep it short as it is current, then fine for me.

```suggestion
In Gardener components, `V(0)` of the root logger (the one that is not "derived" from any other logger; usually constructed on component startup) maps to `info` level, `V(1)` means `debug` level.
```

Reply:
Yes, that's valid. I explained it in the document now.